src/class-model.ts

```diff
@@ -251,23 +278,56 @@ export function register<Instance, Klass extends { new (...args: any[]): Instanc
   });
 
   // create the MST type for not-readonly versions of this using the views and actions extracted from the class
-  klass.mstType = mstTypes
+  let mstType = mstTypes
     .model(klass.name, mstPropsFromQuickProps(klass.properties))
     .views((self) => bindToSelf(self, mstViews))
     .actions((self) => bindToSelf(self, mstActions));
 
   if (Object.keys(mstVolatiles).length > 0) {
     // define the volatile properties in one shot by running any passed initializers
-    (klass as any).mstType = (klass as any).mstType.volatile((self: any) => initializeVolatiles({}, self, mstVolatiles));
+    mstType = mstType.volatile((self: any) => initializeVolatiles({}, self, mstVolatiles));
+  }
+
+  klass.snapshottedViews = metadatas.filter((metadata) => metadata.type == "snapshotted-view") as SnapshottedViewMetadata[];
+  if (klass.snapshottedViews.length > 0) {
+    mstType = mstType
+      .props({ __snapshottedViewsEpoch: mstTypes.optional(mstTypes.frozen(), 0) })
```

Comment:
Curious why this is a prop and not a volatile? Also, why a frozen and not a number?

If there's good reason for it, I think it'd be worth it to drop in a comment as to why.

Reply:
Good questions! I should definitely add a comment to explain this.

> why a frozen and not a number?

It doesn't need to be, I'm just dumb 🤦 -- I'll change this from `mstTypes.frozen` -> `mstTypes.number` 👍

> Curious why this is a prop and not a volatile?

This is used to track changes to a `@snapshottedView` when none of its model's properties changed. This scenario is an anti-pattern, but exists in the place that inspired `@snapshottedView` in the first place (`ModelAction.options`).

Here's the test that the `__snapshottedViewsEpoch` is meant to resolve:

```ts
test("an observable instance emits a patch when the view value changes", () => {
  const observableArray = observable.array<string>([]);

  @register
  class MyViewExample extends ClassModel({ key: types.identifier, name: types.string }) {
    @snapshottedView()
    get arrayLength() {
      return observableArray.length;
    }
  }

  const fn = jest.fn();
  const instance = MyViewExample.create({ key: "1", name: "Test" });
  onPatch(instance, fn);

  runInAction(() => {
    observableArray.push("a");
  });
  expect(fn).toMatchSnapshot();
});
```

If `__snapshottedViewsEpoch` was volatile, the `fn` wouldn't be called because changes to volatile properties don't emit patches.

Reply:
So I guess the underlying reason for this is that we want a patch stream on volatiles? I framed it that way since you'd have a patch if `observableArray` were instead a `types.array(types.string)` prop.

I'm not a huge fan of changing things this way, as it makes the mental model for MST even more complex. Won't block on it, but perhaps we should have considered modeling a snapshotted view as a regular `view` accompanied by a specially-named property that gets set on a reaction to the view.

Reply:
We chatted IRL and I'll summarize.

Definitely could get things working with props, but then we have to think about / deal with two other things and how they could impact the system:

1. Patches will now have large-ish blobs in them
2. We'll retain the "view snapshots" in properties, possibly doubling up the memory of those views.

Not easy to reason about, so we'll be pragmatic and take the slight complexity increase to get the ball rolling on the urgent needs!